### PR TITLE
Homework25: Vladislav Savchuk 

### PR DESCRIPTION
**Pull request by VladislavSavchuk**
# Testing web site: Contact List App

Link: https://thinking-tester-contact-list.herokuapp.com/.

- Written tests for a web site using the Page Object Model pattern. 
- Structured the code in such a way that it was easy to maintain, change and work with it
- I put in separate modules locators for each page, pages, constants for tests and tests themselves. 
- Added logging to tests

### CODE REVIEW

File: homework/homework25/pages/add_contact_page.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+"""This module contains the class for the add contact page"""
+
+from selenium.webdriver.support import expected_conditions as EC
+from homework.homework25.page_locators.add_contact_page import (
+    AddContactLocators)
```
Comment (Stml89) [inline homework/homework25/pages/add_contact_page.py]:
```suggestion
from homework.homework25.page_locators.add_contact_page import AddContactLocators
```

File: homework/homework25/pages/add_contact_page.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+from homework.homework25.page_locators.update_contact_page import (
```
Comment (Stml89) [inline homework/homework25/pages/add_contact_page.py]: same

File: homework/homework25/pages/base_page.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        Raises:
+            TimeoutException: If the element is not found or
+            interactable within the timeout period.
+        """
+        element = self.find_element(locator)
+        if element:
+            element.clear()
+            element.send_keys(text)
+        else:
+            print(f"Element not found: {locator}")
```
Comment (Stml89) [inline homework/homework25/pages/base_page.py]: return `False` or `raise assert`

File: homework/homework25/pages/add_contact_page.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        'birthdate': AddContactLocators.birthdate_input,
+        'email': AddContactLocators.email_input,
+        'phone': AddContactLocators.phone_input,
+        'street1': AddContactLocators.street1_input,
+        'street2': AddContactLocators.street2_input,
+        'city': AddContactLocators.city_input,
+        'state': AddContactLocators.state_input,
+        'postal_code': AddContactLocators.postal_code_input,
+        'country': AddContactLocators.country_input,
+    }
```
Comment (Stml89) [inline homework/homework25/pages/add_contact_page.py]: I'm not too fond of this approach.
Better to import locators or define them in the class

Comment (Stml89) [reply homework/homework25/pages/add_contact_page.py]: fix it in other modules as well

**Review by Stml89 — CHANGES_REQUESTED** [review]

File: homework/homework25/pages/add_contact_page.py (hunk 1 above)
Comment (VladislavSavchuk) [reply homework/homework25/pages/add_contact_page.py]: changed import transfer in all modules

File: homework/homework25/pages/add_contact_page.py (hunk 2 above)
Comment (VladislavSavchuk) [reply homework/homework25/pages/add_contact_page.py]: changed import transfer in all modules

File: homework/homework25/pages/base_page.py (hunk 3 above)
Comment (VladislavSavchuk) [reply homework/homework25/pages/base_page.py]: return False instead print

File: homework/homework25/pages/add_contact_page.py (hunk 4 above)
Comment (VladislavSavchuk) [reply homework/homework25/pages/add_contact_page.py]: define locators in the class in add contact and edit contact modules

**Comment by VladislavSavchuk** [comment]
@Stml89, ready for review

File: homework/homework25/pages/edit_contact_page.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        self.edit_contact_btn = EditContactLocators.edit_contact_btn
+        self.delete_contact_btn = EditContactLocators.delete_contact_btn
+        self.submit_btn = EditContactLocators.submit_btn
+        self.return_to_contacts_btn = (
+            EditContactLocators.return_to_contacts_btn)
+        self.contact_form_fields = {
+            'email': EditContactLocators.email_form,
+            'city': EditContactLocators.city_form,
+            'postal_code': EditContactLocators.postal_code_form
+        }
```
Comment (Stml89) [inline homework/homework25/pages/edit_contact_page.py]: Why do need this?
Use locators as you do on lines #51-52

File: homework/homework25/pages/login_page.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+class LoginPage(BasePage):
+    """
+    Class representing the login page of the application.
+    Inherits from BasePage to provide common page functionalities.
+    """
+    input_email = LoginPageLocators.input_email
+    input_password = LoginPageLocators.input_password
+    submit_btn = LoginPageLocators.submit_btn
+    contact_header = LoginPageLocators.contacts_header
```
Comment (Stml89) [inline homework/homework25/pages/login_page.py]: the same

**Review by Stml89 — CHANGES_REQUESTED** [review]

**Comment by VladislavSavchuk** [comment]
@Stml89, привет) исправлено! 
Удалил папку с модулями локаторов и теперь определяю их внутри каждого класса, кажется так будет проще их отслеживать. Немного поправил методы, которые раньше использовали эти модули.

**Review by Stml89 — APPROVED** [review]

**Pull request merged**